pkg/common/commands/resetoffset/controller/reconciler.go

```diff
@@ -0,0 +1,195 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/Shopify/sarama"
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/reconciler"
+
+	kafkav1alpha1 "knative.dev/eventing-kafka/pkg/apis/kafka/v1alpha1"
+	"knative.dev/eventing-kafka/pkg/client/injection/reconciler/kafka/v1alpha1/resetoffset"
+	kafkalisters "knative.dev/eventing-kafka/pkg/client/listers/kafka/v1alpha1"
+	"knative.dev/eventing-kafka/pkg/common/commands/resetoffset/refmappers"
+	kafkasarama "knative.dev/eventing-kafka/pkg/common/kafka/sarama"
+)
+
+// Component For Sarama Config
+const Component = "reset-offset-controller"
+
+var (
+	_ resetoffset.Interface = (*Reconciler)(nil) // Verify Reconciler Implements Interface
+	_ resetoffset.Finalizer = (*Reconciler)(nil) // Verify Reconciler Implements Finalizer
+)
+
+// Reconciler Implements controller.Reconciler for ResetOffset Resources
+type Reconciler struct {
+	kafkaBrokers      []string
+	saramaConfig      *sarama.Config
+	resetoffsetLister kafkalisters.ResetOffsetLister
+	refMapper         refmappers.ResetOffsetRefMapper
+}
+
+// ReconcileKind implements the Reconciler Interface and is responsible for performing Offset repositioning.
+func (r *Reconciler) ReconcileKind(ctx context.Context, resetOffset *kafkav1alpha1.ResetOffset) reconciler.Event {
+
+	// Get The Logger From Context
+	logger := logging.FromContext(ctx).Desugar()
+	logger.Debug("<==========  START RESET-OFFSET RECONCILIATION  ==========>")
+
+	// Ignore Previously Initiated ResetOffset Instances
+	if resetOffset.Status.IsInitiated() || resetOffset.Status.IsCompleted() {
+		logger.Debug("Skipping reconciliation of previously executed ResetOffset instance")
+		return reconciler.NewEvent(corev1.EventTypeNormal, ResetOffsetSkipped.String(), "Skipped previously executed ResetOffset")
+	}
+
+	// TODO - This functionality should be handled by the Webhook (Admissions / Defaulting) - Verify & Remove once that is in place.
+	//      - Note: The "consolidated" and "distributed" KafkaChannel implementations have similar logic - possibly investigate why?
+	//      - Note: If we remove here and rely on Webhook then we are effectively requiring all users of the Controller to setup a Webhook?
+	// Set Defaults & Verify The ResetOffset Is Valid
+	resetOffset.SetDefaults(ctx)
+	if err := resetOffset.Validate(ctx); err != nil {
+		logger.Error("Received invalid ResetOffset", zap.Error(err))
+		return fmt.Errorf("received invalid ResetOffset: %v", err)
```

Comment:
FYI you don't need SetDefaults() too, I've saw it this morning while reading some of the code generated code and pkg/reconciler runtime code that `SetDefaults()` is invoked before `ReconcileKind` is invoked

Reply:
OK good to know - I tested it without this logic and it works ok assuming you're reconciling a valid CRD.  I will remove it for now.  I had also created an Issue to address this in the other controllers as Matthias suggested (Issue #657).